[WIP] Added auto-reload of device plugins when configmap is changed.

gageorsburn opened the pull request:
Fixes issue #49 

*Important Notes*
* This removes environment variables from the bridge plugin all together. 
* Must have service account for namespace or be deployed in the kube-system namespace.
* Adds in k8s.io/client-go as a dependency. (if you guys know a better way without adding in the dependency please let me know as I suck with golang still)

gageorsburn committed:
Added auto-reload of device plugins when configmap is changed.

ovirt-infra commented:
Hello contributor, thanks for submitting a PR for this project! 

I am the bot who triggers "standard-CI" builds for this project.
As a security measure, I will not run automated tests on PRs that are not from white-listed contributors.

In order to allow automated tests to run, please ask one of the project maintainers to review the code and then do one of the following:

1. Type `ci test please` on this PR to trigger automated tests for it.
2. Type `ci add to whitelist` on this PR to trigger automated tests for it and also add you to the contributor white-list so that your future PRs will be tested automatically.
3. If you are planning to contribute to more than one project, maybe it's better to ask them to add you to the project organization, so you'll be able to run tests for all the organization's projects.

gageorsburn commented:
Yeah I'm pretty bad with dependency management in Go. Let me know what I need to do to fix the 6,000 file diff lol.

Actual changed files:

	modified:   Gopkg.toml
	modified:   cmd/network/bridge/bridge.go
	modified:   pkg/network/bridge/lister.go

phoracek commented on cmd/network/bridge/bridge.go:
```diff
@@ -2,20 +2,18 @@ package main
-	if !bridgesListDefined {
-		glog.Exit("BRIDGES environment variable must be set in format BRIDGE[,BRIDGE[...]]")
-	}
+	// _, bridgesListDefined := os.LookupEnv(bridge.BridgesListEnvironmentVariable)
```
Please remove the unused code completely.

phoracek commented on pkg/network/bridge/lister.go:
```diff
@@ -22,19 +25,36 @@ func (bl BridgeLister) GetResourceNamespace() string {
-	for _, bridgeName := range bridges {
-		if len(bridgeName) > maxBridgeNameLength {
-			glog.Fatalf("Bridge name (%s) cannot be longer than %d characters", bridgeName, maxBridgeNameLength)
+		configmap, err := clientset.CoreV1().ConfigMaps("kube-system").Get("device-plugin-network-bridge", metav1.GetOptions{})
```
Instead of polling, we should use monitoring using list/watch. A tiny example: https://stackoverflow.com/questions/40975307/how-to-watch-events-on-a-kubernetes-service-using-its-go-client

gageorsburn replied:
Thanks, didn't know that existed, will implement.

phoracek reviewed (changes requested):
Thanks for the PR!

gageorsburn committed:
Changed monitoring method from polling to list and watch.

gageorsburn commented:
@phoracek Changes made. Thanks for the suggestion!

phoracek commented on pkg/network/bridge/lister.go:
```diff
@@ -15,26 +20,60 @@ const (
 	maxBridgeNameLength = 12
 )
 
+func GetBridgeList(configMap *v1.ConfigMap) dpm.PluginNameList {
```
This should be a private function, starting with lower case.

phoracek commented on pkg/network/bridge/lister.go:
```diff
@@ -15,26 +20,60 @@ const (
-	var plugins = make(dpm.PluginNameList, 0)
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		panic(err.Error())
```
Please use glog.Fatalf with error included.

phoracek commented on pkg/network/bridge/lister.go:
```diff
@@ -15,26 +20,60 @@ const (
-			glog.Fatalf("Bridge name (%s) cannot be longer than %d characters", bridgeName, maxBridgeNameLength)
-		}
-		plugins = append(plugins, bridgeName)
+	configmap, err := clientset.CoreV1().ConfigMaps("kube-system").Get("device-plugin-network-bridge", metav1.GetOptions{})
```
Currently it is not a requirement to create this DP in kube-system namespace. I think it would make sense to do that. Would you please update README to expose that?

gageorsburn replied:
@phoracek Updated the manifest and read me,

phoracek commented on pkg/network/bridge/lister.go:
```diff
@@ -15,26 +20,60 @@ const (
 	}
 
-	pluginListCh <- plugins
+	pluginListCh <- GetBridgeList(configmap)
```
I believe that the informer passes you all existing items to AddFunc when it starts.

We should just initially pass an empty list of bridges here and then read existing/wait for a new one in the informer.

gageorsburn replied:
@phoracek The informer will only watch for updates so if the plugin pod got restarted no bridges would be discovered until the config map is updated again. Line 60 passes the existing bridges in the config map to the channel for the initial discovery.

phoracek replied:
The current solution is raceful, the map may change between this line and startup of the informer.

If you add AddFunc callback to the informer, you would get the initial map (maybe you need to set an extra argument or I am totally wrong).

Also if someone changes the map by removing it and creating again, this would not work either.

gageorsburn replied:
@phoracek You are correct sir. I have pushed that change. Let me know if there's anything else you spot.

phoracek commented on pkg/network/bridge/lister.go:
```diff
@@ -15,26 +20,60 @@ const (
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			cm := newObj.(*v1.ConfigMap)
+			if cm.GetName() == "device-plugin-network-bridge" {
+				glog.Infoln(cm.Data)
```
Please add more info to the log call, something like:

```
glog.V(3).Infof("Config map was updated: %v", cm.Data)
```

gageorsburn committed:
Updated bridge readme and manifest to reflect namespace requirement. Added better logging to lister and fixed function accessiblity.

gageorsburn committed:
Added AddFunc in.

yuvalif commented on pkg/network/bridge/lister.go:
```diff
@@ -15,26 +20,61 @@ const (
-	pluginListCh <- plugins
+	watchlist := cache.NewListWatchFromClient(clientset.CoreV1().RESTClient(), string(v1.ResourceConfigMaps), v1.NamespaceAll, fields.Everything())
+
+	_, informer := cache.NewInformer(watchlist, &v1.ConfigMap{}, 0, cache.ResourceEventHandlerFuncs{
```
do we want to handle map deletion?

phoracek replied:
In case a user removes original map and creates a new one, it would be beneficial not to remove it. If a user wants to expose zero bridges, just set empty list. I think it may work better like this, not sure.

yuvalif replied:
sure. but I think we should still support that.
i mean we will have 2 events, once deleting the map, and another one creating it.

phoracek replied:
Ok, I'm not against it. Let's add delete callback too.

phoracek replied:
@gageorsburn ^^

gageorsburn replied:
@phoracek Yup doing it now actually. Sorry I had to put out some unrelated dayjob fires.

phoracek replied:
@gageorsburn no rush :)

gageorsburn replied:
@phoracek @yuvalif  Delete callback is implemented now. I really like how it flows now actually.

yuvalif commented on pkg/network/bridge/lister.go:
```diff
@@ -22,19 +25,36 @@ func (bl BridgeLister) GetResourceNamespace() string {
+			if len(bridgeName) > maxBridgeNameLength {
+				glog.Fatalf("Bridge name (%s) cannot be longer than %d characters", bridgeName, maxBridgeNameLength)
+			}
+			plugins = append(plugins, bridgeName)
```
do we need to handle the case where a bridge is removed from the list?

yuvalif commented on Gopkg.toml:
```diff
@@ -34,6 +34,14 @@
 
 [[constraint]]
-  version = "~v1.10"
+  version = "v1.10"
```
why not support all 1.10 minor version?

gageorsburn replied:
@yuvalif  Had some issues with client-go syncing up. I couldn't get it to compile with all 1.10 minor version.

gageorsburn committed:
Added DeleteFunc callback.

phoracek commented on docs/README.bridge.md:
```diff
@@ -8,6 +8,8 @@ This device plugin allows user to attach pods to existing bridges on nodes.
 
 ## Usage
 
+_The plugin must be run in the kube-system namespace._
```
@gageorsburn we need to update configmap configuration in this README too, it does not specify namespace.

gageorsburn replied:
@phoracek Good catch. Namespaces updated in both places.

phoracek replied:
@gageorsburn still failing:

```
I0815 12:29:20.189786    3268 reflector.go:240] Listing and watching *v1.ConfigMap from github.com/kubevirt/kubernetes-device-plugins/pkg/network/bridge/lister.go:85
E0815 12:29:20.192274    3268 reflector.go:205] github.com/kubevirt/kubernetes-device-plugins/pkg/network/bridge/lister.go:85: Failed to list *v1.ConfigMap: configmaps is forbidden: User "system:serviceaccount:kube-system:default" cannot list configmaps in the namespace "kube-system"
```

Are you sure kube-system does not need any RBAC configuration? Maybe there is a different trick there, if not, we can drop the Namespace requirement and create a new Role.

gageorsburn replied:
@phoracek That's really odd. I just tested with a fresh Minikube install and no issues.

Log: https://pastebin.com/MFL9HwEL

In what conditions are you trying to run it? I'm curious at this point and would like to better understand the RBAC nuances.

phoracek replied:
@gageorsburn does Minikube have RBAC enabled?

Build DP image and sync it with local cluster:

```
make build-network-bridge docker-build-network-bridge cluster-down cluster-up cluster-sync-network-bridge
```

Create daemon set with the DP:

```
./cluster/kubectl.sh create -f ds.yml
```

ds.yml:

```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  labels:
    name: device-plugin-network-bridge
  name: device-plugin-network-bridge
  namespace: kube-system
spec:
  selector:
    matchLabels:
      name: device-plugin-network-bridge
  template:
    metadata:
      labels:
        name: device-plugin-network-bridge
    spec:
      hostNetwork: true
      hostPID: true
      containers:
      - name: device-plugin-network-bridge
        image: registry:5000/device-plugin-network-bridge
        securityContext:
          privileged: true
        volumeMounts:
          - name: var-run
            mountPath: /var/run
          - name: device-plugin
            mountPath: /var/lib/kubelet/device-plugins
      volumes:
        - name: var-run
          hostPath:
            path: /var/run
        - name: device-plugin
          hostPath:
            path: /var/lib/kubelet/device-plugins
```

And finally read logs using `./cluster/kubectl.sh logs`.

gageorsburn replied:
It did yeah, I explicitly passed in the RBAC flag. You can see it in the log.

```
 gageorsburn@mbp  ~/go/src/github.com/kubevirt/kubernetes-device-plugins   autoconfigmap  minikube start --feature-gates=DevicePlugins=true --extra-config=apiserver.authorization-mode=RBAC
```

I'll try that and see what happens.

gageorsburn replied:
Yeah that's strange. I'm not sure why minikube with RBAC enabled is different than the Makefile cluster. On minikube the default namespace would give me this same error but kube-system did not. Would you rather us create a new service account or just give kube-system default read-only permissions?

`create rolebinding default-view --clusterrole=view --serviceaccount=kube-system:default --namespace=kube-system`

@phoracek

phoracek commented on pkg/network/bridge/lister.go:
```diff
@@ -1,40 +1,88 @@
 	}
 
-	pluginListCh <- plugins
+	watchlist := cache.NewListWatchFromClient(clientset.CoreV1().RESTClient(), string(v1.ResourceConfigMaps), v1.NamespaceAll, fields.Everything())
```
This fails with:

```
I0814 12:13:58.790363    2450 reflector.go:240] Listing and watching *v1.ConfigMap from github.com/kubevirt/kubernetes-device-plugins/pkg/network/bridge/lister.go:85
E0814 12:13:58.791881    2450 reflector.go:205] github.com/kubevirt/kubernetes-device-plugins/pkg/network/bridge/lister.go:85: Failed to list *v1.ConfigMap: configmaps is forbidden: User "system:serviceaccount:kube-system:default" cannot list configmaps at the cluster scope
```

We should list only the kube-system namespace, not NamespaceAll.

gageorsburn committed:
Updated namespace configurations.

gageorsburn committed:
Switching to other PC. Haven't tested yet. Adding RBAC.
